### PR TITLE
Max element

### PR DESCRIPTION
**Pull request by gress2**

### CODE REVIEW

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp:171]: can we wrap the two shifts in static_cast?

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp:228]: Why wouldn't we just  return `_shift_towards_msb(read, first_position())`?

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp:238]: Same idea here. It ends up only saving us a single instruction probably. 

Essentially we no longer need the `& mask` but at the cost of losing the original position of the bits.

File: include/bit_algorithm_details.hpp
Comment (gress2) [reply include/bit_algorithm_details.hpp:228]: I think there might be confusion about what the function is doing. If we have 01011101 with first pointed at the 3rd bit (0 indexed) and last pointed at one past the last bit i.e. ^01011^101 and we want to do a padded read by padding the bits outside the iterators with 0s, we should end up with 01011000. 01011101 << 3 = 11101000

Comment (bkille) [reply include/bit_algorithm_details.hpp:228]: Yea so I  get that, but I guess maybe I should've made the same point as I did in the comment below. 

If we don't care about the position of the bit, why cant we just shift the original word instead? It would save us from having to perform the mask operation.

Comment (gress2) [reply include/bit_algorithm_details.hpp:228]: I guess because then you can run into the issue of arbitrarily adding 0's or 1's to your word where it's not appropriate.

Comment (bkille) [reply include/bit_algorithm_details.hpp:228]: yes but on this specific line you know that you are filling with 0s, so the only behavioral difference is the position. I suppose in `max`, you need to know the position. So you would have to increment the result of `tzcnt` with `first.position()` which is probably more expensive than the mask. 

I guess my point is we should only use `_padded_read` when we need to know the position of the bits as well. Maybe we could have an overloaded implementation of this such that we shift off the out of range bits instead of mask them? Again this only matters in the `bv=0` case, but it is still worth saving an operation IMO.

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: I think the `template <class It>` should specify which type of iterator can be passed to it. For example in this case, we cannot safely use this function on an Input Iterator

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: I think the template <class It> should specify which type of iterator can be passed to it. For example in this case, we cannot safely use this function on an Input Iterator

File: include/max_element.hpp
Comment (bkille) [inline include/max_element.hpp]: If we're dealing with full words, I think we should use iterators over words. Calling `std::advance(cursor, num_digits)` is `O(num_digits)` for anything less powerful than an RAI. We have an implementation of `bit::advance` I believe, but even then, we end up using the modulo operator to determine how many times to advance the word iterator. I'm sure a lot of this can be done at compile time, since `num_digits` is known then, but I'd want to see evidence of that given that we call it so many times.

File: include/max_element.hpp
Comment (bkille) [inline include/max_element.hpp]: Won't `cursor.position()` always be `0` here? We align it on line 47, and then only add whole words to it.

File: include/bit_algorithm_details.hpp
Comment (bkille) [reply include/bit_algorithm_details.hpp:228]: Actually, it looks like [there is already arithmetic anyways](https://github.com/vreverdy/bit-algorithms/pull/14/files/c43a8e751697b24f13bc1bda17874460f51b7004#diff-7c6afd9d6651f72f61b9032e6f7569e0R39):

39: `return cursor + (first_set_bit_position - cursor.position());`

If instead we simply shifted off the out of range bits, we would then have
39: `return cursor + first_set_bit_position;`


And for [the case in the final word](https://github.com/vreverdy/bit-algorithms/pull/14/files/c43a8e751697b24f13bc1bda17874460f51b7004#diff-7c6afd9d6651f72f61b9032e6f7569e0R69), it makes more sense to mask the end off to keep the position the same, otherwise we end up with more arithmetic.

69: `return cursor + (last_set_bit_position - cursor.position());`

would end up being 

69: `return`cursor + last_set_bit_position - (digits - last.position())`

File: include/max_element.hpp
Comment (bkille) [inline include/max_element.hpp]: If we use whole word iterators, this is another line that we could get rid of I think.

File: include/bit_algorithm_details.hpp
Comment (bkille) [reply include/bit_algorithm_details.hpp:238]: And for the version that does end up masking, the canonical way to do this is 

`mask = ((1<<(last_pos-first_pos))-1) << (first_pos);` 
as opposed to 
`mask = ((-1) << first_pos) & ((-1) >> (digits-last_pos));`

which [saves a whopping **ONE** instruction!  ](https://godbolt.org/z/tPnaeD)

File: src/tests/max_element.hpp
Comment (bkille) [inline src/tests/max_element.hpp:27]: I think that since we have the framework and tools in place to, we should be testing each algorithm using the template product so that we can test on each of the applicable iterator types as well.

File: include/bit_algorithm_details.hpp
Comment (bkille) [reply include/bit_algorithm_details.hpp:238]: [I just tested all 3 out.](https://godbolt.org/z/69LOSr) Shifting input word to shave off the edges uses the least instructions.

File: include/max_element.hpp
Comment (bkille) [inline include/max_element.hpp]: Header should be aligned

File: include/max_element.hpp
Comment (bkille) [inline include/max_element.hpp:59]: Do you know if calling the constructor on every iteration is also optimized? 

If we have cursor as a word iterator, we still need to increment it at every step, so I'm not worried about that. But is there any overhead to calling the `bit_iterator` constructor on every iteration?

I think the rest of my comments on this PR are very nitpicky and don't offer much performance gain, but I am interested in this one.

File: include/bit_algorithm_details.hpp
Comment (gress2) [reply include/bit_algorithm_details.hpp:171]: are you sure we need to? was curious so I tried it https://godbolt.org/z/fr8mZ6

**Comment by gress2** [comment]
I think I addressed everything in my last commit. Feel free to unresolve if needed

File: include/max_element.hpp
Comment (bkille) [inline include/max_element.hpp]: So here, we are doing 2 conditional statements for each iteration of the for loop:
`!(last.base() == word_cursor)` and `_is_one_past_last_bit(last, word_cursor)`

I think it can be done in 1 conditional statement:

`std::next(word_cursor, _is_aligned_lsb(last)) != last.base()` assuming that the function result is recognized as a constant by the compiler. The statement serves the same purpose, go until there are no more full words.

File: include/max_element.hpp
Comment (bkille) [inline include/max_element.hpp:61]: I think we can avoid creating another bit iterator. The only remaining conditions depend on the alignment of the `last` iterator and the only time we use `cursor` is to access the base word anyways. Not worth holding up the PR, though.

**Review by bkille — APPROVED** [review]

**Pull request merged**
